src/WidgetMessaging.js

```diff
@@ -109,9 +115,54 @@ export default class WidgetMessaging {
 
     start() {
         this.fromWidget.addEndpoint(this.widgetId, this.widgetUrl);
+        this.fromWidget.addListener("get_openid", this._openIdHandlerRef);
     }
 
     stop() {
         this.fromWidget.removeEndpoint(this.widgetId, this.widgetUrl);
+        this.fromWidget.removeListener("get_openid", this._openIdHandlerRef);
+    }
+
+    async _onOpenIdRequest(ev, rawEv) {
+        if (ev.widgetId !== this.widgetId) return; // not interesting
+
+        const settings = SettingsStore.getValue("widgetOpenIDPermissions");
+        if (settings.blacklist && settings.blacklist.includes(this.widgetId)) {
+            this.fromWidget.sendResponse(rawEv, {state: "blocked"});
+            return;
+        }
+        if (settings.whitelist && settings.whitelist.includes(this.widgetId)) {
+            const responseBody = {state: "allowed"};
+            const credentials = await MatrixClientPeg.get().getOpenIdToken();
+            Object.assign(responseBody, credentials);
+            this.fromWidget.sendResponse(rawEv, responseBody);
+            return;
+        }
+
+        // Confirm that we received the request
+        this.fromWidget.sendResponse(rawEv, {state: "request"});
+
+        // Actually ask for permission to send the user's data
+        Modal.createTrackedDialog("OpenID widget permissions", '',
+            WidgetOpenIDPermissionsDialog, {
+                widgetUrl: this.widgetUrl,
+                widgetId: this.widgetId,
+
+                onFinished: async (confirm) => {
+                    const responseBody = {success: confirm};
+                    if (confirm) {
+                        const credentials = await MatrixClientPeg.get().getOpenIdToken();
+                        Object.assign(responseBody, credentials);
```

Comment:
Sorry for not being more familiar with the widget messaging API, but what's the reason for having two response formats here? E.g. one with just credentials for sendResponse for black/whitelisted widgets, and here {credentials} and an action? Could this also not be a response?

Reply:
This is different because it is a different direction. The other one is a response to a `fromWidget` request whereas this is a new request over `toWidget`.